### PR TITLE
Refactor Servlet API and JDBC plugins to use a MethodHandles-based advice dispatching approach

### PR DESCRIPTION
**Pull request by felixbarny**
First step of elastic/apm-agent-java#1085

### CODE REVIEW

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/FilterChainInstrumentation.java [hunk 1]
```diff
@@ -64,9 +66,25 @@ public FilterChainInstrumentation(ElasticApmTracer tracer) {
...
     }
 
-    @Override
-    public Class<?> getAdviceClass() {
-        return ServletApiAdvice.class;
+    @Nullable
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    private static Transaction onEnterServletService(@Advice.Origin Class<?> clazz,
+                                                     @Advice.Argument(0) ServletRequest servletRequest) throws Throwable {
+        return (Transaction) MethodHandleDispatcher
```
Comment (felixbarny) [inline apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/FilterChainInstrumentation.java]: Referencing agent types in advice methods might not be possible anymore once we load the agent from an isolated class loader.

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/AbstractServletInstrumentation.java [hunk 2]
```diff
@@ -40,6 +42,11 @@
         return Collections.singleton(SERVLET_API);
     }
 
+    @Override
+    public List<String> helpers() throws Exception {
+        return PackageScanner.getClassNames(getClass().getPackage().getName());
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/AbstractServletInstrumentation.java:47]: Same- since this would be a reference for future usages, I think it's better to extract the `@RegisterMethodHandle` methods to dedicated classes in dedicated packages. It's easier to make mistakes if we mix instrumentations and advices with those. It also makes sense that packages reflect the complicated class loading contexts.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/HelperClassManager.java [hunk 3]
```diff
@@ -271,7 +290,148 @@ private synchronized T loadAndReferenceHelper(Class<?> classOfTargetClassLoader)
...
+                    } catch (Exception e) {
+                        logger.error("Exception while trying to register method handles for {}", name);
+                        if (e instanceof InvocationTargetException) {
+                            Throwable targetException = ((InvocationTargetException) e).getTargetException();
+                            if (targetException instanceof Exception) {
+                                throw (Exception) targetException;
+                            }
+                        }
+                        throw e;
+                    }
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/bci/HelperClassManager.java:400]: Lazily injecting the helper classes has the downside that we wouldn't be able to leverage an upcoming feature in Byte Buddy. This feature would inject an `INVOKEDYNAMIC` instruction to call the helper method: https://github.com/raphw/byte-buddy/issues/830#issuecomment-603955721
This avoids the overhead of having to look up the method handle from a map.

However, there are a few things we can do to make this safer:
* We already only load classes where we checked with Byte Buddy matchers that they contain `@RegisterMethodHandle`-annotated methods.
* Call `getMethods` instead of `getDeclaredMethods`: avoids resolving types of all parameters of non-public methods. So we could make methods with problematic signatures private or package-private.
* Using the upcoming `bootstrap` method, I think we'd be able to get the method handles without having to use reflection. We'd need to do some testing but I suspect we'll need a fallback for Java 7 that does not rely on `INVOKEDYNAMIC`.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/util/PackageScanner.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                }
+                classNames.addAll(listClassNames(basePackage, basePath));
+            }
+        }
+        return classNames;
+    }
+
+    private static List<String> listClassNames(final String basePackage, final Path basePath) throws IOException {
+        final List<String> classNames = new ArrayList<>();
+        Files.walkFileTree(basePath, new SimpleFileVisitor<Path>() {
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/util/PackageScanner.java:82]: This was also a new discovery for me 🙂

File: apm-agent-core/src/test/java/co/elastic/apm/agent/util/PackageScannerTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    public void getDoesNotContainTestClasses() throws Exception {
+        assertThat(PackageScanner.getClassNames(getClass().getPackageName()))
+            .doesNotContain(getClass().getName());
+    }
+
+    @Test
+    public void testScanJar() throws Exception {
+        assertThat(PackageScanner.getClassNames(ByteBuddy.class.getPackageName()))
+            .contains(ByteBuddy.class.getName());
+        assertThat(PackageScanner.getClassNames(ByteBuddy.class.getPackageName()))
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-core/src/test/java/co/elastic/apm/agent/util/PackageScannerTest.java:50]: Yes, it tests avoiding `FileSystemAlreadyExistsException`. Will rename the method.

File: apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/JdbcInstrumentation.java [hunk 6]
```diff
@@ -25,31 +25,19 @@
...
-        synchronized (JdbcInstrumentation.class) {
-            if (jdbcHelperManager == null) {
-                jdbcHelperManager = HelperClassManager.ForSingleClassLoader.of(tracer,
-                    "co.elastic.apm.agent.jdbc.helper.JdbcHelperImpl",
-                    "co.elastic.apm.agent.jdbc.helper.JdbcHelperImpl$1");
-            }
-        }
+    @Override
+    public List<String> helpers() throws Exception {
+        return PackageScanner.getClassNames(getClass().getPackage().getName());
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/JdbcInstrumentation.java:40]: My vision for an ideal situation (which requires some changes in Byte Buddy) is that advices can be written without having to manually get a `MethodHandle`. When we load all methods from the helper class loader, you can even implement an instrumentation with just one file, without having to worry about which classes are able to access application classes.
After all plugins have been refactored, we can pull this method up, making it the default behavior to include all classes.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/HelperClassManager.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/bci/HelperClassManager.java:404]: > Lazily injecting the helper classes has the downside that we wouldn't be able to leverage an upcoming feature in Byte Buddy

Is it something we should rely on? And won't we have to do some adjustments anyhow?

> Call getMethods instead of getDeclaredMethods: avoids resolving types of all parameters of non-public methods. So we could make methods with problematic signatures private or package-private.

The nature of these problems is that they are very setup-dependant, so there are not "problematic signatures" per se. In most cases it will work, but somewhere (eg some module-dependency systems) these side effects may cause a lookup of a type that is not yet available.

Bottom line- it is safer and if the effort is not big, I think it worth it (as the effort to debug even a single related problem could be bigger). However, if the effort is substantial, or if you think it will require a rework effort in the foreseeable future - let's merge as is and rethink later if needed. It's really your call, either way is fine for me.

File: apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/JdbcInstrumentation.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (eyalkoren) [reply apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/JdbcInstrumentation.java:40]: > you can even implement an instrumentation with just one file

In general, a separation will reduce the chances for errors. It's always better if you have a class exactly once in your classpath. Consider such an instrumentation class with a static field, or some initialization code run in the instrumentation constructor. What can actually be seen in runtime from these annotated method is very difficult to understand when just looking at such file.

Again- not a blocker.

**Comment by felixbarny** [comment]
superseded by #1230
